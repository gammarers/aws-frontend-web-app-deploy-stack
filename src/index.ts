import { SecureBucket, SecureBucketType } from '@gammarers/aws-secure-bucket';
import { SecureFrontendWebAppCloudFrontDistribution } from '@gammarers/aws-secure-frontend-web-app-cloudfront-distribution';
import * as cdk from 'aws-cdk-lib';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as alias from 'aws-cdk-lib/aws-route53-targets';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import { Construct } from 'constructs';

export interface CloudFrontOptions {
  readonly priceClass?: cloudfront.PriceClass;
}

export interface FrontendWebAppDeployStackProps extends cdk.StackProps {
  readonly domainName: string;
  readonly hostedZoneId: string;
  readonly originBucketName: string;
  readonly logBucketArn: string;
  readonly deploySourceAssetPath: string;
  readonly cloudFrontOptions?: CloudFrontOptions;
}

export class FrontendWebAppDeployStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: FrontendWebAppDeployStackProps) {
    super(scope, id, props);

    // ðŸ‘‡Create Secure Cloud Front Origin Bucket
    const originBucket = new SecureBucket(this, 'SecureCloudFrontOriginBucket', {
      bucketName: props.originBucketName,
      bucketType: SecureBucketType.CLOUD_FRONT_ORIGIN,
    });

    // ðŸ‘‡Get Hosted Zone.
    const hostedZone = route53.HostedZone.fromHostedZoneAttributes(this, 'HostedZone', {
      hostedZoneId: props.hostedZoneId,
      zoneName: props.domainName,
    });

    // ðŸ‘‡Create Certificate with DNS validate.
    const certificate = new acm.Certificate(this, 'Certificate', {
      domainName: props.domainName,
      validation: acm.CertificateValidation.fromDns(hostedZone),
    });

    // ðŸ‘‡Secure Frontend Web App CloudFront Distribution.
    const distribution = new SecureFrontendWebAppCloudFrontDistribution(this, 'SecureFrontendWebAppCloudFrontDistribution', {
      comment: 'frontend web app distribution.',
      accessLogBucket: s3.Bucket.fromBucketArn(this, 'LogBucket', props.logBucketArn),
      certificate: certificate,
      domainName: props.domainName,
      originBucket: originBucket,
      priceClass: props.cloudFrontOptions?.priceClass,
    });

    // ðŸ‘‡Route 53 DNS (Alias)
    new route53.ARecord(this, 'DNSARecord', {
      zone: hostedZone,
      recordName: props.domainName,
      target: route53.RecordTarget.fromAlias(new alias.CloudFrontTarget(distribution)),
      ttl: cdk.Duration.seconds(300),
    });

    // ðŸ‘‡Deploy to Bucket with CloudFront Cache invalidate
    new s3deploy.BucketDeployment(this, 'DeployWithInvalidation', {
      sources: [s3deploy.Source.asset(props.deploySourceAssetPath)],
      destinationBucket: originBucket,
      prune: true,
      distribution,
      distributionPaths: ['/*'],
    });
  }
}
