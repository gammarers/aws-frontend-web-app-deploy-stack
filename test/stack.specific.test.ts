import * as cdk from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import { FrontendWebAppDeployStack } from '../src';

describe('FrontendWebAppDeployStack Testing', () => {

  const app = new cdk.App();
  const stack = new FrontendWebAppDeployStack(app, 'FrontendWebAppDeployStack', {
    env: { account: '012345678901', region: 'us-east-1' },
    domainName: 'example.com',
    hostedZoneId: 'Z0000000000000000000Q',
    originBucketName: 'frontend-web-app-example-origin-bucket',
    deploySourceAssetPath: 'website/',
    logBucketArn: 'arn:aws:s3:::frontend-web-app-example-access-log-bucket',
    cloudFrontOptions: {
      priceClass: cloudfront.PriceClass.PRICE_CLASS_100,
    },
  });
  const template = Template.fromStack(stack);

  it('Has CloudFront Distribution', () => {
    template.hasResourceProperties('AWS::CloudFront::Distribution', Match.objectLike({
      DistributionConfig: {
        Aliases: [
          'example.com',
        ],
      },
    }));
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});