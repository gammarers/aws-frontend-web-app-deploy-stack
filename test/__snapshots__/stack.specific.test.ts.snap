// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontendWebAppDeployStack Testing Should match snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "example.com",
        "DomainValidationOptions": [
          {
            "DomainName": "example.com",
            "HostedZoneId": "Z0000000000000000000Q",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "FrontendWebAppDeployStack/Certificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-012345678901-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "DeployWithInvalidationAwsCliLayerDEDD5787",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-012345678901-us-east-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-012345678901-us-east-1/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SecureCloudFrontOriginBucket152632A1",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "SecureCloudFrontOriginBucket152632A1",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DNSARecord1B875602": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "SecureFrontendWebAppCloudFrontDistributionFBE118EA",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "Z0000000000000000000Q",
        "Name": "example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "DeployWithInvalidationAwsCliLayerDEDD5787": {
      "Properties": {
        "Content": {
          "S3Bucket": "cdk-hnb659fds-assets-012345678901-us-east-1",
          "S3Key": "HASH.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "DeployWithInvalidationCustomResourceE3FF7455": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "SecureCloudFrontOriginBucket152632A1",
        },
        "DistributionId": {
          "Ref": "SecureFrontendWebAppCloudFrontDistributionFBE118EA",
        },
        "DistributionPaths": [
          "/*",
        ],
        "OutputObjectKeys": true,
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          "cdk-hnb659fds-assets-012345678901-us-east-1",
        ],
        "SourceObjectKeys": [
          "HASH.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "ResponseHeadersPolicy13DBF9E0": {
      "Properties": {
        "ResponseHeadersPolicyConfig": {
          "Comment": "A default policy",
          "CustomHeadersConfig": {
            "Items": [
              {
                "Header": "Cache-Control",
                "Override": true,
                "Value": "no-cache, no-store, private",
              },
              {
                "Header": "pragma",
                "Override": true,
                "Value": "no-cache",
              },
            ],
          },
          "Name": "FrontendWebAppDeployStackResponseHeadersPolicy70756A2F",
          "SecurityHeadersConfig": {
            "ContentTypeOptions": {
              "Override": true,
            },
            "FrameOptions": {
              "FrameOption": "SAMEORIGIN",
              "Override": true,
            },
            "ReferrerPolicy": {
              "Override": true,
              "ReferrerPolicy": "strict-origin-when-cross-origin",
            },
            "StrictTransportSecurity": {
              "AccessControlMaxAgeSec": 31536000,
              "IncludeSubdomains": true,
              "Override": true,
              "Preload": true,
            },
            "XSSProtection": {
              "ModeBlock": true,
              "Override": true,
              "Protection": true,
            },
          },
        },
      },
      "Type": "AWS::CloudFront::ResponseHeadersPolicy",
    },
    "SecureCloudFrontOriginBucket152632A1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "Private",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": "frontend-web-app-example-origin-bucket",
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "BucketOwnerEnforced",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:d794ec86",
            "Value": "true",
          },
        ],
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SecureCloudFrontOriginBucketPolicy8A0340C6": {
      "Properties": {
        "Bucket": {
          "Ref": "SecureCloudFrontOriginBucket152632A1",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SecureCloudFrontOriginBucket152632A1",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "SecureCloudFrontOriginBucket152632A1",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Condition": {
                "StringEquals": {
                  "AWS:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/",
                        {
                          "Ref": "SecureFrontendWebAppCloudFrontDistributionFBE118EA",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudfront.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "SecureCloudFrontOriginBucket152632A1",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "SecureFrontendWebAppCloudFrontDistributionFBE118EA": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "example.com",
          ],
          "Comment": "frontend web app distribution.",
          "CustomErrorResponses": [
            {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            {
              "ErrorCachingMinTTL": 300,
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "CachedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "Compress": true,
            "ResponseHeadersPolicyId": {
              "Ref": "ResponseHeadersPolicy13DBF9E0",
            },
            "TargetOriginId": "FrontendWebAppDeployStackSecureFrontendWebAppCloudFrontDistributionOrigin18253E717",
            "ViewerProtocolPolicy": "https-only",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2and3",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": "frontend-web-app-example-access-log-bucket.s3.us-east-1.amazonaws.com",
            "Prefix": "example.com/",
          },
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "SecureCloudFrontOriginBucket152632A1",
                  "RegionalDomainName",
                ],
              },
              "Id": "FrontendWebAppDeployStackSecureFrontendWebAppCloudFrontDistributionOrigin18253E717",
              "OriginAccessControlId": {
                "Fn::GetAtt": [
                  "SecureFrontendWebAppCloudFrontDistributionOrigin1S3OriginAccessControl09AD5FFF",
                  "Id",
                ],
              },
              "S3OriginConfig": {
                "OriginAccessIdentity": "",
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "Certificate4E7ABB08",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "SecureFrontendWebAppCloudFrontDistributionOrigin1S3OriginAccessControl09AD5FFF": {
      "Properties": {
        "OriginAccessControlConfig": {
          "Name": "FrontendWebAppDeployStackSecOrigin1S3OriginAccessControl239764ED",
          "OriginAccessControlOriginType": "s3",
          "SigningBehavior": "always",
          "SigningProtocol": "sigv4",
        },
      },
      "Type": "AWS::CloudFront::OriginAccessControl",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
